tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/LockFreeExponentiallyDecayingReservoir.java

```diff
@@ -0,0 +1,193 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.tritium.metrics.registry;
+
+import com.codahale.metrics.Clock;
+import com.codahale.metrics.Reservoir;
+import com.codahale.metrics.Snapshot;
+import com.codahale.metrics.WeightedSnapshot;
+import com.codahale.metrics.WeightedSnapshot.WeightedSample;
+import com.google.common.annotations.Beta;
+import com.palantir.logsafe.Preconditions;
+import java.time.Duration;
+import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+
+/**
+ * {@link LockFreeExponentiallyDecayingReservoir} is based closely on the codahale
+ * <a href="https://github.com/dropwizard/metrics/blob/0313a104bf785e87d7d14a18a82026225304c402/metrics-core/src/main/java/com/codahale/metrics/ExponentiallyDecayingReservoir.java">
+ * ExponentiallyDecayingReservoir.java</a>, however it provides looser guarantees (updates which race rescale may be
+ * lost) while completely avoiding locks.
```

Comment:
If I'm reading correctly, in the case of racing rescales, the winner would have its new state installed and the losers would perform a volatile read of the latest state (which under extremely small rescale threshold may actually be newer than the former winner) and return that state, which is then updated. We might want to consider enforcing a lower bound for a reasonable rescale threshold may be specified (say 1 second which even seems low?).

Reply:
Your analysis is correct. I’m not sure I’d want to strictly enforce a minimum interval because there are so many variables, perhaps stronger documentation describing the potential pitfalls would be sufficient?

I think losing a cas and updating a newer-than-expected state object is roughly equivalent to the calling thread being descheduled for a moment as it calls into reservoir.update.